Deprecate memory corruption and sensitive data tests for Android and iOS (#3506)
OWASP/owasp-mstg: knowledge/android/MASVS-STORAGE/MASTG-KNOW-0051.md

title: Process Memory
---

Like any regular modern-day computer, all Android applications use memory to perform normal computational operations. Therefore, it is no surprise that sensitive operations are sometimes performed within process memory. For this reason, it is important to dispose of sensitive data from process memory as quickly as possible once it has been processed.

> The investigation of an application's memory can be done from memory dumps, and from analyzing the memory in real time via a debugger.

Author: ```suggestion
The investigation of an application's memory can be done from memory dumps, and from analyzing the memory in real time via a debugger or a binary instrumentation framework (see @MASTG-TECH-0044 and @MASTG-TECH-0031).
```

title: Process Memory
---

> Like any regular modern-day computer, all Android applications use memory to perform normal computational operations. Therefore, it is no surprise that sensitive operations are sometimes performed within process memory. For this reason, it is important to dispose of sensitive data from process memory as quickly as possible once it has been processed.

Reviewer: ```suggestion
Many applications handle sensitive information, provided either by the user, the operating system or the backend. This data is often available in-memory while the user is using the application, as it is needed for the application to function. When the application no longer actively needs the sensitive data, it should dispose of it immediately.
```
The original phrasing was a bit weird (computer vs app, 'no surprise')

Using non-primitive mutable types like `StringBuffer` and `StringBuilder` may be acceptable, but it's indicative and requires care. Types like `StringBuffer` are used to modify content (which is what you want to do). To access such a type's value, however, you would use the `toString` method, which would create an immutable copy of the data. There are several ways to use these data types without creating an immutable copy, but they require more effort than using a primitive array. Safe memory management is one benefit of using types like `StringBuffer` , but this can be a two-edged sword. If you try to modify the content of one of these types and the copy exceeds the buffer capacity, the buffer size will automatically increase. The buffer content may be copied to a different location, leaving the old content without a reference use to overwrite it.

> 

Author: ```suggestion

Support for true in-place destruction is rare, and even when implemented, it cannot remove other copies that may exist in heap memory. This is an inherent limitation of Java and Android memory management for in-heap secrets. Examples include:

- Overwriting the byte array returned by `secretKey.getEncoded()` [does not clear the internal key](https://bugs.openjdk.org/browse/JDK-6263419), because `getEncoded()` returns a copy. Zeroing that copy has no effect on the bytes held by the key object.
- `SecretKeySpec` does not provide a functional `destroy()` implementation, and this [gap has been tracked in the JDK](https://bugs.openjdk.org/browse/JDK-8160206). Calling `destroy()` on a `SecretKey` backed by `SecretKeySpec` therefore does not guarantee that the key material is wiped.
- RSA keys handled outside the Android Keystore are backed by immutable `BigInteger` objects, which cannot be cleared once created, leaving their data in memory until garbage collected.
- User-provided data (such as credentials, social security numbers, or credit card information) entered through `EditText` is delivered via the `Editable` interface. If your app does not supply a custom `Editable.Factory`, the data will likely remain in memory longer than necessary. The default `Editable` implementation, `SpannableStringBuilder`, has the same retention issues as Java’s `StringBuilder` and `StringBuffer`, leaving sensitive content in memory until garbage collected.
```

> 

Reviewer 2: I removed this paragraph, there is no destroy method for [SecretKeySpec](https://developer.android.com/reference/javax/crypto/spec/SecretKeySpec) and the sentence "See the sections below for the proper way to remove a SecretKey from memory." was redundant.

Author: I brought back these paragraphs but in a better shape as I think the info is relevant and provides context (added more links for additional context). See lines 21-26.

Support for true in-place destruction is rare, and even when implemented, it cannot remove other copies that may exist in heap memory. This is an inherent limitation of Java and Android memory management for in-heap secrets. Examples include:

- Overwriting the byte array returned by `secretKey.getEncoded()` [does not clear the internal key](https://bugs.openjdk.org/browse/JDK-6263419), because `getEncoded()` returns a copy. Zeroing that copy has no effect on the bytes held by the key object.
- `SecretKeySpec` does not provide a functional `destroy()` implementation, and this [gap has been tracked in the JDK](https://bugs.openjdk.org/browse/JDK-8160206). Calling `destroy()` on a `SecretKey` backed by `SecretKeySpec` therefore does not guarantee that the key material is wiped.
- RSA keys handled outside the Android Keystore are backed by immutable `BigInteger` objects, which cannot be cleared once created, leaving their data in memory until garbage collected.

> - User-provided data (such as credentials, social security numbers, or credit card information) entered through `EditText` is delivered via the `Editable` interface. If your app does not supply a custom `Editable.Factory`, the data will likely remain in memory longer than necessary. The default `Editable` implementation, `SpannableStringBuilder`, has the same retention issues as Java’s `StringBuilder` and `StringBuffer`, leaving sensitive content in memory until garbage collected.

Author: ```suggestion
- User-provided data (such as credentials, social security numbers, or credit card information) entered through `EditText` is delivered via the `Editable` interface. If your app does not supply a custom `Editable.Factory`, the data will likely remain in memory longer than necessary. The default `Editable` implementation, `SpannableStringBuilder`, has the same retention issues as Java's `StringBuilder` and `StringBuffer`, leaving sensitive content in memory until garbage collected.
```

> **Why doesn't the OWASP MASTG have tests for this anymore?**
>
> The tests for sensitive data in process memory were removed from the OWASP MASTG because they provide limited practical value in assessing real-world risks and are difficult to execute reliably. Memory inspection typically requires privileged access, such as root or jailbreak, which falls outside the threat model for most production environments. Such access already compromises the device's overall security, making memory-level findings less meaningful in isolation.
>
> Moreover, results from memory dumps are highly variable and prone to false positives or negatives depending on timing, garbage collection, and system-level optimizations. These inconsistencies make the test unsuitable as a standardized or repeatable security control.
>

> > Instead of relying on runtime memory inspection, OWASP now emphasizes preventing these issues through secure design and development practices. Frameworks like the [OWASP Software Assurance Maturity Model (SAMM)](https://owaspsamm.org/model/) and [NIST Special Publication (SP) 800-218, Secure Software Development Framework (SSDF)](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-218.pdf) guide developers to handle sensitive data securely in memory, minimize exposure windows, and use safe APIs for cryptography and data storage. The related OWASP MASVS categories on secure data storage (@MASVS-STORAGE), secure communication (@MASVS-NETWORK), and secure authentication and session management (@MASVS-AUTH) already cover the necessary safeguards to prevent sensitive information from being inadvertently exposed at runtime.

Reviewer 2: ```suggestion
> Instead of relying on runtime memory inspection, OWASP now emphasizes preventing these issues through secure design and development practices. Frameworks like the [OWASP Software Assurance Maturity Model (SAMM)](https://owaspsamm.org/model/) and [NIST Special Publication (SP) 800-218, Secure Software Development Framework (SSDF)](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-218.pdf) guide developers to handle sensitive data securely in memory, minimize exposure windows, and use safe APIs for cryptography and data storage. The related OWASP MASVS categories on secure data storage (@MASVS-STORAGE), secure communication (@MASVS-NETWORK), and secure authentication and session management (@MASVS-AUTH) already cover the necessary safeguards to prevent sensitive information from being inadvertently exposed at runtime. To support this, developers should follow the [best practices](https://developer.android.com/privacy-and-security/risks/use-of-native-code) for the use of native code in Android apps, as defined by Android.
```
